mne/preprocessing/xdawn.py

```diff
@@ -318,63 +417,82 @@ class Xdawn(_Xdawn, ContainsMixin):
-        if self.correct_overlap == 'auto':
-            self.correct_overlap = _check_overlapp(epochs)
-
-        events, epochs_data, event_id = self._preproc_epochs(epochs)
-        self._set_signal_cov(epochs_data)
-        self._fit_xdawn(epochs_data, events[:, -1], event_id, events,
-                        epochs.tmin, epochs.tmax, epochs.info['sfreq'],
-                        epochs.baseline)
-
-        self.ch_names = epochs.ch_names
-        self.exclude = np.arange(self.n_components, len(self.ch_names))
-        self.event_id = event_id
-        self._tmin = epochs.tmin
-        self._get_evokeds(epochs.info, tmin=epochs.tmin)
+        # Check data
+        if not isinstance(epochs, _BaseEpochs):
+            raise ValueError('epochs must be an Epochs object.')
+        X = epochs.get_data()
+        X = X[:, _pick_data_channels(epochs.info), :]
+        y = epochs.events[:, 2] if y is None else y
+
+        # Check that no baseline was applied with correct overlap
+        correct_overlap = self.correct_overlap
+        if correct_overlap == 'auto':
+            # Events are overlapped if the minimal inter-stimulus
+            # interval is smaller than the time window.
+            isi = np.diff(epochs.events[:, 0])
+            window = int((epochs.tmax - epochs.tmin) *
+                         epochs.info['sfreq'])
+            correct_overlap = isi.min() < window
+
+        if epochs.baseline and correct_overlap is True:
+            raise ValueError('Cannot apply correct_overlap if epochs'
+                             ' were baselined.')
+
+        events, tmin, sfreq = None, 0., 1.
+        if correct_overlap is True:
+            events = epochs.events
+            tmin = epochs.tmin
+            sfreq = epochs.info['sfreq']
+
+        # Note: In this original version of Xdawn we compute and keep all
+        # components. The selection comes at transform().
+        n_components = X.shape[1]
+
+        # Main fitting function
+        filters, patterns, evokeds = _fit_xdawn(
+            X, y,  n_components=n_components, reg=self.reg,
+            signal_cov=self.signal_cov, events=events, tmin=tmin, sfreq=sfreq)
+
+        # Re-order filters and patterns according to event_id
+        filters = filters.reshape(-1, n_components, filters.shape[-1])
+        patterns = patterns.reshape(-1, n_components, patterns.shape[-1])
+        self.filters_, self.patterns_, self.evokeds_ = dict(), dict(), dict()
+        idx = np.argsort([value for _, value in iteritems(epochs.event_id)])
+        for eid, this_filter, this_pattern, this_evo in zip(
+                epochs.event_id, filters[idx], patterns[idx], evokeds[idx]):
+            self.filters_[eid] = this_filter.T
+            self.patterns_[eid] = this_pattern.T
+            n_events = len(epochs[eid])
+            evoked = EvokedArray(this_evo, epochs.info, tmin=epochs.tmin,
+                                 comment=eid, nave=n_events)
+            self.evokeds_[eid] = evoked
```

Comment:
This is the old `Xdawn` API. I'm for deprecating this class in favour of `XdawnTransformer`.